{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e088d74-28cc-432e-b676-0a70e3243823",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Users must check whether A2utils.py file has benn downloaded\n",
    "import A2utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dca4e3d8-299b-4ed8-9a1a-8e0040faf777",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from typing import Tuple\n",
    "from Bio import AlignIO\n",
    "from Bio import Align\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b61df61c-bf3b-40a9-8bac-9aca3384b832",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_msa(filepath: str):\n",
    "    msa = AlignIO.read(filepath, \"fasta\")\n",
    "    return AlignInfo.SummaryInfo(msa.alignment)\n",
    "\n",
    "def load_sequences(filepath: str):\n",
    "    from Bio import SeqIO\n",
    "    seqs = []\n",
    "    for seq in SeqIO.parse(filepath, \"fasta\"):\n",
    "        seqs.append((seq.name, str(seq.seq)))\n",
    "    return seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "620d8244-55de-4913-8112-590d81919f02",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# We implement the UPGMA for phylo tree construction\n",
    "# We assign sequence name as label under each node, and each node is an unique integers.\n",
    "# Edge length is added under len attribute.\n",
    "def grab_leaves(tree, internal_node, leaf_nodes_lst):\n",
    "    descendants_inter = set(nx.descendants(tree, internal_node))\n",
    "    leaves_for_internal_nodes = list(descendants_inter & set(leaf_nodes_lst))\n",
    "    return leaves_for_internal_nodes\n",
    "\n",
    "def build_evo_tree(sequences: Tuple[str, str], msa: AlignInfo.SummaryInfo) -> nx.DiGraph:\n",
    "    \"\"\"\n",
    "    Performs evolutionary tree building using input sequences and/or an input msa.\n",
    "    \"\"\"\n",
    "    # YOUR CODE HERE\n",
    "    # calculate the distance matrix\n",
    "    aligner = Align.PairwiseAligner()\n",
    "    try:\n",
    "        aligner.substitution_matrix = Bio.Align.substitution_matrices.load('BLOSUM80')\n",
    "    except:\n",
    "        aligner.substitution_matrix = Bio.Align.substitution_matrices.load('blosum80')\n",
    "    n = len(seqs)\n",
    "    labels = [seqs[i][0] for i in range(n)]\n",
    "    dist_mat = pd.DataFrame(np.full(shape = (n, n), fill_value = ''))\n",
    "    for i in range(n):\n",
    "        for j in range(n):\n",
    "            if i == j:\n",
    "                dist_mat.at[i, j] = np.nan  # if idx is the same, put nan first\n",
    "            else:\n",
    "                if dist_mat.loc[i, j] != '':\n",
    "                    dist_mat.at[i, j] = dist_mat.loc[j, i]  # if we've calculated the distance for that pair, just replace its value (no need to calculate again)\n",
    "                else:\n",
    "                    costs = aligner.align(msa.alignment[i], msa.alignment[j]).score\n",
    "                    dist_mat.at[i, j] = costs\n",
    "\n",
    "    max_score = dist_mat.max().max()\n",
    "    dist_mat = 1 - dist_mat/max_score  # score normalization\n",
    "    dist_mat = dist_mat.fillna(0)      # fill the nan as zero\n",
    "\n",
    "    # First adding all leaf nodes inside the phylogenic tree\n",
    "    leaf_dct = dict(zip([i for i in range(len(labels))], labels)) # create a dict for leaf nodes, the structure is number : strain_name\n",
    "    tree = nx.DiGraph()\n",
    "    for i in range(len(leaf_dct)):\n",
    "        tree.add_node(i, label = leaf_dct[i])  # First ingest all the nodes\n",
    "\n",
    "    # inter_name = [] # internal nodes, same as leaf_dct; number : strain_name\n",
    "    innode_name = n  # set the idx for internal node\n",
    "\n",
    "    # start to build the tree\n",
    "    while n > 1:\n",
    "        # find the minimum distance inside the tree\n",
    "        min_dist = dist_mat[dist_mat != 0].min().min()  # zero distance means two selected nodes from col and row are the same, we don't want to pick it up.\n",
    "        cols = list(dist_mat.columns)\n",
    "        # find the first pair with the minimum distance\n",
    "        min_i, min_j = cols[np.where(dist_mat == min_dist)[0][0]],  cols[np.where(dist_mat == min_dist)[1][0]]\n",
    "        \n",
    "        # two nodes are leaf nodes\n",
    "        if leaf_dct.get(min_i) and leaf_dct.get(min_j):\n",
    "            tree.add_edge(innode_name, min_i, weight = min_dist/2)\n",
    "            tree.add_edge(innode_name, min_j, weight = min_dist/2)\n",
    "            # inter_name.append(innode_name)\n",
    "            node1 = min_i; node2 = min_j\n",
    "\n",
    "        # other cases. Writing else directly is because I want to make it clear.\n",
    "        else:\n",
    "            # one in leaf, and the other is internal\n",
    "            if leaf_dct.get(min_i) is None and leaf_dct.get(min_j):\n",
    "                leaf_idx = min_j; inter_idx = min_i\n",
    "                tree.add_edge(innode_name, leaf_idx, weight = min_dist/2)\n",
    "                descendants_inter = set(nx.descendants(tree, inter_idx)); leaves_for_inter = list(descendants_inter & set(leaf_dct.keys()))\n",
    "                tree.add_edge(innode_name, inter_idx, weight = min_dist/2 - nx.shortest_path_length(G = tree, source = inter_idx, target = leaves_for_inter[0]) )\n",
    "                node1 = leaf_idx; node2 = inter_idx\n",
    "\n",
    "            # one in leaf, and the other is internal\n",
    "            elif leaf_dct.get(min_j) is None and leaf_dct.get(min_i):\n",
    "                leaf_idx = min_i; inter_idx = min_j\n",
    "                tree.add_edge(innode_name, leaf_idx, weight = min_dist/2)\n",
    "                descendants_inter = set(nx.descendants(tree, inter_idx)); leaves_for_inter = list(descendants_inter & set(leaf_dct.keys()))\n",
    "                tree.add_edge(innode_name, inter_idx, weight = min_dist/2 - nx.shortest_path_length(G = tree, source = inter_idx, target = leaves_for_inter[0]) )\n",
    "                node1 = leaf_idx; node2 = inter_idx\n",
    "\n",
    "          # Two are all internal nodes\n",
    "            else:\n",
    "                leaves_for_inter1 = grab_leaves(tree, min_i, list(leaf_dct.keys()))\n",
    "                leaves_for_inter2 = grab_leaves(tree, min_j, list(leaf_dct.keys()))\n",
    "                tree.add_edge(innode_name, min_i, weight = min_dist/2 - nx.shortest_path_length(G = tree, source = min_i, target = leaves_for_inter1[0]) )\n",
    "                tree.add_edge(innode_name, min_j, weight = min_dist/2 - nx.shortest_path_length(G = tree, source = min_j, target = leaves_for_inter2[0]) )\n",
    "                node1 = min_i; node2 = min_j\n",
    "\n",
    "        # update distance matrix\n",
    "        new_distance = []\n",
    "        for k in dist_mat.index:\n",
    "            # calculate the distance of nodes left in the distance matrix\n",
    "            if k != node1 and k != node2:\n",
    "                new_dist = (dist_mat.loc[node1, k] + dist_mat.loc[node2, k])/2\n",
    "                new_distance.append(new_dist)\n",
    "        \n",
    "        # We drop two nodes selected in this round and append the newly-formed internal node inside the distance matrix\n",
    "        dist_mat = dist_mat.drop([node1, node2], axis = 1).drop([node1, node2], axis = 0)\n",
    "        dist_mat[innode_name] = 0; dist_mat.loc[innode_name] = 0\n",
    "        dist_mat[innode_name] = new_distance + [0]; dist_mat.loc[innode_name] = new_distance + [0]\n",
    "\n",
    "        n = len(dist_mat); innode_name += 1  # update n and update the internal node index\n",
    "\n",
    "    return tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38eb3597-a993-44fc-a972-f83ac264bc3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We can load the MSA.fasta and fasta file into the function.\n",
    "\n",
    "seqs = load_sequences(EBOLA_SEQS_FILEPATH)\n",
    "msa = load_msa(EBOLA_MSA_FILEPATH)\n",
    "T = build_evo_tree(seqs, msa)\n",
    "A2utils.draw_evo_tree(T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67db78a2-b4de-43e9-b954-41911989e2d0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
